jvm/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseCodegenStages.scala

```diff
@@ -121,13 +122,19 @@ case class ColumnarCollapseCodegenStages(
     plan match {
       case p if !supportTransform(p) =>
         new ColumnarInputAdapter(insertWholeStageTransformer(p))
+      // The children of BaseJoinExec should do transformer separately.
+      case j if j.isInstanceOf[BaseJoinExec] =>
+        j.withNewChildren(j.children.map(c => insertWholeStageTransformer(c)))
```

Comment:
The final Spark plan tree seems not right. Maybe something's missing here:
```
j.withNewChildren(j.children.map(c =>
    new ColumnarInputAdapter(insertWholeStageTransformer(c))))
```

Reply:
Fixed